### PR TITLE
[RTM] Fixed {{toggle_view}} insert tag

### PR DESCRIPTION
**Pull request by Toflar**
Same issue as mentioned multiple times: When a page is cached (in case of Contao 3.5) there's no environment information anymore. That's why every fragment that wants to be included using ESI (in Contao 4) needs to make sure it works 100% without any dependencies.
I've added the `request` query parameter to pass on the original request to the ESI request string and while I was on it, also passed on the `pageId` parameter.
Again, there might be a lot of other use cases and we just cannot fix them all with the software architecture we currently have. I'm working on it and these small PR's are tiny step-by-step fixes so that at least the core works fine (if an extension dev needs another param, we might provide it until the proper solution was implemented).

### CODE REVIEW

File: src/Resources/contao/library/Contao/InsertTags.php [hunk 1]
```diff
@@ -562,8 +565,19 @@ protected function doReplace($strBuffer, $blnCache)
 
 				// Mobile/desktop toggle (see #6469)
 				case 'toggle_view':
-					$strUrl = ampersand(\Environment::get('request'));
+					$strRequest = \Environment::get('request');
+
+					// ESI request
+					if (preg_match('/^' . preg_quote(ltrim(\System::getContainer()->getParameter('fragment.path'), '/'), '/') . '/', $strRequest))
```
Comment (leofeyer) [inline src/Resources/contao/library/Contao/InsertTags.php]: What exactly does this do?

Comment (Toflar) [inline src/Resources/contao/library/Contao/InsertTags.php]: `Environment::get('request')` now returns `_fragment?......&insertTag=....&toggle_view=desktop` which is not the link you want to show ;-) You want to add this to the original master request which is why this is passed on as query param to the ESI request.

File: src/Resources/contao/library/Contao/InsertTags.php [hunk 2]
```diff
@@ -120,7 +120,10 @@ protected function doReplace($strBuffer, $blnCache)
 				{
 					/** @var FragmentHandler $fragmentHandler */
 					$fragmentHandler = \System::getContainer()->get('fragment.handler');
-					$strBuffer .= $fragmentHandler->render(new ControllerReference('contao.controller.insert_tags:renderAction', ['insertTag' => '{{' . $strTag . '}}']), 'esi');
+					$strBuffer .= $fragmentHandler->render(new ControllerReference('contao.controller.insert_tags:renderAction',
+						['insertTag' => '{{' . $strTag . '}}'],
+						['pageId' => $objPage->id, 'request' => \Environment::get('request')]
```
Comment (aschempp) [inline src/Resources/contao/library/Contao/InsertTags.php]: I suggest `request` should be called `referer`

Comment (leofeyer) [inline src/Resources/contao/library/Contao/InsertTags.php]: @Toflar What about @aschempp's suggestion?

Comment (Toflar) [inline src/Resources/contao/library/Contao/InsertTags.php]: I think `request` is better. `Referrer` is an HTTP header which does not apply here. It's a request query attribute. It doesn't really matter, I just found `request` to be the better name which is why I chose it.

File: src/Resources/contao/library/Contao/InsertTags.php [hunk 3]
```diff
@@ -562,8 +565,19 @@ protected function doReplace($strBuffer, $blnCache)
...
+					if (preg_match('/^' . preg_quote(ltrim(\System::getContainer()->getParameter('fragment.path'), '/'), '/') . '/', $strRequest))
+					{
+						$request = \System::getContainer()->get('request_stack')->getCurrentRequest();
+						$strRequest = $request->query->get('request');
+					}
+
+					$strUrl = ampersand($strRequest);
 					$strGlue = (strpos($strUrl, '?') === false) ? '?' : '&amp;';
+					
+					\System::loadLanguageFile('default');
```
Comment (leofeyer) [inline src/Resources/contao/library/Contao/InsertTags.php]: Why did you add `\System::loadLanguageFile('default');` here?

Comment (Toflar) [inline src/Resources/contao/library/Contao/InsertTags.php]: Otherwise the button is not translated because nobody was loading the file :)

**Comment by leofeyer** [comment]
Thank you @Toflar.
